# 8. Volumes Aggregation from SynthSeg Outputs

## Obiettivo
Aggregare i volumi cerebrali e i QC scores da SynthSeg per tutti i dataset disponibili nel NAS corrente.

## Dataset Processati (pipeline SynthSeg)
1. **OASIS-2**
2. **OASIS-3**
3. **ADNI**
4. **IXI**
5. **PPMI**
6. **SRPBS**
7. **AABC**
8. **FOMO**

## Nota su RF
- Il dataset **RF** e` usato per analisi clinico-radiologiche/concordanza multimediale, non come coorte normativa SynthSeg di training.
- RF ha quindi un notebook dedicato di esplorazione separato dalla pipeline di aggregazione SynthSeg.

## Output
Per ogni dataset viene creato un CSV in `/data/volumes/` contenente:
- **metriche volumetriche** con prefisso `vol_`
- **QC scores** con prefisso `qc_`
- **subject_id** e, se presente, **session_id**

## Struttura SynthSeg
```
dataset/derivatives/synthseg/
└── {subject_id}/
    ├── volumes.csv          # Volumi regionali
    ├── qc_scores.csv        # Quality control scores
    ├── segmentation.nii.gz  # Maschera di segmentazione
    └── resampled.nii.gz     # T1w resampleata
```

## 1. Setup

In [1]:
import os
import pandas as pd
from pathlib import Path

## 2. Define Paths

Percorsi alle cartelle `derivatives/synthseg/` per ciascun dataset nel NAS corrente.

**Note**:
- Le path sono parametrizzate con `NAS_ROOT` per ridurre hardcode e drift tra mount point.
- OASIS viene gestito in due rami separati (`OASIS2` e `OASIS3`).
- RF non e` incluso qui per scelta metodologica (pipeline separata non normativa).

In [9]:
# Paths to SynthSeg derivatives folders (allineati al NAS corrente)
NAS_ROOT = Path("/mnt/NAS-Progetti/BrainAtrophy/Normative Model")

def first_existing_path(*candidates: Path) -> Path:
    """Return first existing path, otherwise keep first candidate as default."""
    for candidate in candidates:
        if candidate.exists():
            return candidate
    return candidates[0]

SYNTHSEG_PATHS = {
    "adni": NAS_ROOT / "ADNI/derivatives/synthseg",
    "ixi": NAS_ROOT / "IXI/derivatives/synthseg",
    "oasis2": first_existing_path(
        NAS_ROOT / "OASIS/OASIS2_BIDS/derivatives/synthseg",
        NAS_ROOT / "OASIS/derivatives/synthseg/OASIS2",
    ),
    "oasis3": first_existing_path(
        NAS_ROOT / "OASIS/OASIS3_BIDS/derivatives/synthseg",
        NAS_ROOT / "OASIS/derivatives/synthseg/OASIS3",
    ),
    "ppmi": first_existing_path(
        NAS_ROOT / "PPMI/nifti/derivatives/synthseg",
        NAS_ROOT / "PPMI/nifti_baseline/derivatives/synthseg",
    ),
    "srpb": NAS_ROOT / "SRPBS/SRPBS_BIDS/derivatives/synthseg",
    "aabc": NAS_ROOT / "AABC/derivatives/synthseg",
    "fomo": NAS_ROOT / "FOMO/derivatives/synthseg",
}

## 3. Function: Find Volumes and QC Files

Questa funzione:
1. Scandisce ricorsivamente la cartella del dataset
2. Cerca directory contenenti sia `volumes.csv` che `qc_scores.csv`
3. Estrae il `subject_id` dal nome della directory
4. Restituisce un DataFrame con i path ai file per ogni soggetto

In [3]:
def find_volumes_and_qc(path):
    """
    Find all volumes.csv and qc_scores.csv files in a dataset directory.

    Args:
        path (str or Path): Path to dataset derivatives/synthseg folder

    Returns:
        pd.DataFrame: DataFrame with columns [subject_id, session_id, volumes, qc]
    """
    path = Path(path)
    records = []

    if not path.exists():
        return pd.DataFrame(columns=["subject_id", "session_id", "volumes", "qc"])

    for root, dirs, files in os.walk(path):
        if "volumes.csv" in files and "qc_scores.csv" in files:
            root_path = Path(root)
            session_id = None
            subject_id = None

            # BIDS-like parsing
            for part in reversed(root_path.parts):
                if session_id is None and part.startswith("ses-"):
                    session_id = part
                if subject_id is None and part.startswith("sub-"):
                    subject_id = part

            # Fallback non-BIDS
            if subject_id is None:
                subject_id = root_path.name

            records.append(
                {
                    "subject_id": subject_id,
                    "session_id": session_id,
                    "volumes": str(root_path / "volumes.csv"),
                    "qc": str(root_path / "qc_scores.csv"),
                }
            )

    return pd.DataFrame(records)

## 4. Function: Aggregate Volumes and QC

Questa funzione:
1. Carica i file `volumes.csv` e `qc_scores.csv` per ogni soggetto
2. Aggiunge prefissi `vol_` e `qc_` alle colonne per distinguerle
3. Merge dei due DataFrame sulla colonna subject
4. Verifica che i subject_id corrispondano
5. Aggiunge il subject_id come colonna
6. Concatena tutti i soggetti in un unico DataFrame

**Output**: Un DataFrame con una riga per soggetto e ~110 colonne (101 volumi + 8 QC + subject_id)

In [10]:
def aggregate_volumes_and_qc(volumes_df):
    """
    Aggregate volumes and QC scores for all subjects.

    Args:
        volumes_df (pd.DataFrame): DataFrame from find_volumes_and_qc()

    Returns:
        pd.DataFrame: Aggregated data with vol_*, qc_*, subject_id/session_id
    """
    if volumes_df.empty:
        return pd.DataFrame(columns=["subject_id"])

    aggregated_data = []

    for _, row in volumes_df.iterrows():
        vol_df = pd.read_csv(row["volumes"]).add_prefix("vol_")
        qc_df = pd.read_csv(row["qc"]).add_prefix("qc_")

        merged_df = pd.merge(vol_df, qc_df, left_on="vol_subject", right_on="qc_subject")

        if not merged_df["vol_subject"].equals(merged_df["qc_subject"]):
            raise ValueError(f"Mismatch vol_subject/qc_subject for {row['subject_id']}")

        merged_df["subject_id"] = row["subject_id"]
        merged_df["session_id"] = row["session_id"]

        merged_df = merged_df.drop(columns=["vol_subject", "qc_subject"])
        if not merged_df.empty:
            aggregated_data.append(merged_df)

    if not aggregated_data:
        return pd.DataFrame(columns=["subject_id"])

    # Drop all-NA columns per chunk before concat to avoid pandas FutureWarning.
    concat_chunks = [chunk.dropna(axis=1, how="all") for chunk in aggregated_data]
    out_df = pd.concat(concat_chunks, ignore_index=True, sort=False)

    if "session_id" not in out_df.columns:
        out_df["session_id"] = pd.NA

    # Se presenti sessioni multiple dello stesso soggetto, conserva una riga per soggetto.
    if out_df["subject_id"].duplicated().any():
        out_df = out_df.sort_values(["subject_id", "session_id"], na_position="last")
        out_df = out_df.drop_duplicates(subset="subject_id", keep="first")

    tail_cols = [c for c in ["session_id", "subject_id"] if c in out_df.columns]
    cols = [c for c in out_df.columns if c not in tail_cols] + tail_cols
    return out_df[cols]

## 5. Process All Datasets

### 5.1 Find volumes and QC files

In [11]:
# Find volumes and QC files for each dataset
print("Finding volumes and QC files...\n")

dataset_volumes = {}
for dataset, ds_path in SYNTHSEG_PATHS.items():
    df = find_volumes_and_qc(ds_path)
    dataset_volumes[dataset] = df
    status = "OK" if len(df) > 0 else "EMPTY/MISSING"
    print(f"{dataset.upper():10s}: {len(df):4d} subjects ({status}) - {ds_path}")

total_subjects = sum(len(df) for df in dataset_volumes.values())
print(f"\nTotal subjects with SynthSeg: {total_subjects}")

Finding volumes and QC files...

ADNI      :  520 subjects (OK) - /mnt/NAS-Progetti/BrainAtrophy/Normative Model/ADNI/derivatives/synthseg
IXI       :  274 subjects (OK) - /mnt/NAS-Progetti/BrainAtrophy/Normative Model/IXI/derivatives/synthseg
OASIS2    :  150 subjects (OK) - /mnt/NAS-Progetti/BrainAtrophy/Normative Model/OASIS/OASIS2_BIDS/derivatives/synthseg
OASIS3    : 1376 subjects (OK) - /mnt/NAS-Progetti/BrainAtrophy/Normative Model/OASIS/OASIS3_BIDS/derivatives/synthseg
PPMI      :  160 subjects (OK) - /mnt/NAS-Progetti/BrainAtrophy/Normative Model/PPMI/nifti/derivatives/synthseg
SRPB      : 1410 subjects (OK) - /mnt/NAS-Progetti/BrainAtrophy/Normative Model/SRPBS/SRPBS_BIDS/derivatives/synthseg
AABC      :  369 subjects (OK) - /mnt/NAS-Progetti/BrainAtrophy/Normative Model/AABC/derivatives/synthseg
FOMO      :  864 subjects (OK) - /mnt/NAS-Progetti/BrainAtrophy/Normative Model/FOMO/derivatives/synthseg

Total subjects with SynthSeg: 5123


### 5.2 Aggregate volumes and QC scores

In [12]:
# Aggregate volumes and QC for each dataset
print("\nAggregating volumes and QC scores...\n")

aggregated = {}
for dataset, vols_df in dataset_volumes.items():
    if vols_df.empty:
        print(f"{dataset.upper():10s}: skipped (no files found)")
        continue
    agg_df = aggregate_volumes_and_qc(vols_df)
    aggregated[dataset] = agg_df
    print(f"{dataset.upper():10s}: {agg_df.shape}")


Aggregating volumes and QC scores...

ADNI      : (520, 111)
IXI       : (274, 111)
OASIS2    : (150, 111)
OASIS3    : (1376, 111)
PPMI      : (160, 111)
SRPB      : (1410, 111)
AABC      : (369, 111)
FOMO      : (708, 111)


### 5.3 Inspect column structure

Verifichiamo la struttura delle colonne per un dataset di esempio.

In [7]:
# Show column structure for ADNI as example (fallback to first available dataset)
example_name = "adni" if "adni" in aggregated else next(iter(aggregated.keys()), None)

if example_name is None:
    print("No aggregated dataset available yet.")
else:
    example_df = aggregated[example_name]
    print(f"\nColumn structure ({example_name.upper()} example):")
    print(f"Total columns: {len(example_df.columns)}\n")

    vol_cols = [col for col in example_df.columns if col.startswith("vol_")]
    print(f"Volume metrics: {len(vol_cols)}")
    print("First 10 volume columns:")
    for col in vol_cols[:10]:
        print(f"  - {col}")

    qc_cols = [col for col in example_df.columns if col.startswith("qc_")]
    print(f"\nQC scores: {len(qc_cols)}")
    for col in qc_cols:
        print(f"  - {col}")


Column structure (ADNI example):
Total columns: 111

Volume metrics: 101
First 10 volume columns:
  - vol_total intracranial
  - vol_left cerebral white matter
  - vol_left cerebral cortex
  - vol_left lateral ventricle
  - vol_left inferior lateral ventricle
  - vol_left cerebellum white matter
  - vol_left cerebellum cortex
  - vol_left thalamus
  - vol_left caudate
  - vol_left putamen

QC scores: 8
  - qc_general white matter
  - qc_general grey matter
  - qc_general csf
  - qc_cerebellum
  - qc_brainstem
  - qc_thalamus
  - qc_putamen+pallidum
  - qc_hippocampus+amygdala


## 6. Save Aggregated Data

Salviamo un CSV per dataset nella cartella `/data/volumes/`.

**IMPORTANTE**: I file vengono salvati con `index=False`, ma il subject_id è stato aggiunto come colonna
prima di impostarlo come index, quindi la colonna `subject_id` sarà l'ultima colonna nel CSV finale.

In [13]:
# Create output directory
output_dir = "/home/mario/Repository/Normal_Alzeihmer/data/volumes/"
os.makedirs(output_dir, exist_ok=True)

print(f"Saving aggregated data to {output_dir}\n")

saved = []
for dataset, df in aggregated.items():
    file_name = f"{dataset}.csv"
    save_df = df.sort_values([c for c in ["subject_id", "session_id"] if c in df.columns])
    save_df.to_csv(output_dir + file_name, index=False)
    saved.append((file_name, len(save_df)))
    print(f"OK Saved: {file_name} ({len(save_df)} subjects)")

print(f"\nOK Total files saved: {len(saved)}")
print(f"OK Total subjects: {sum(n for _, n in saved)}")

Saving aggregated data to /home/mario/Repository/Normal_Alzeihmer/data/volumes/

OK Saved: adni.csv (520 subjects)
OK Saved: ixi.csv (274 subjects)
OK Saved: oasis2.csv (150 subjects)
OK Saved: oasis3.csv (1376 subjects)
OK Saved: ppmi.csv (160 subjects)
OK Saved: srpb.csv (1410 subjects)
OK Saved: aabc.csv (369 subjects)
OK Saved: fomo.csv (708 subjects)

OK Total files saved: 8
OK Total subjects: 4967


## 7. Summary Statistics

Calcoliamo alcune statistiche di base per verificare i dati.

In [ ]:
# Summary statistics for a key region (hippocampus)
print("\nSummary statistics for hippocampal volume (left + right):\n")

for name, df in aggregated.items():
    if "vol_left hippocampus" not in df.columns or "vol_right hippocampus" not in df.columns:
        print(f"{name.upper()}: hippocampus columns missing, skip")
        continue

    total_hipp = df["vol_left hippocampus"] + df["vol_right hippocampus"]

    print(f"{name.upper()}:")
    print(f"  Mean: {total_hipp.mean():.1f} mm^3")
    print(f"  Std:  {total_hipp.std():.1f} mm^3")
    print(f"  Range: [{total_hipp.min():.1f}, {total_hipp.max():.1f}] mm^3")
    print()

## Output Files

I seguenti file CSV sono stati creati in `/data/volumes/`:

1. **oasis2.csv** - 150 soggetti × 110 colonne
2. **oasis3.csv** - 1,376 soggetti × 110 colonne
3. **adni.csv** - 520 soggetti × 110 colonne
4. **ixi.csv** - 346 soggetti × 110 colonne
5. **ppmi.csv** - 160 soggetti × 110 colonne
6. **srpb.csv** - 1,410 soggetti × 110 colonne

**Totale: 3,962 soggetti**

## Struttura delle Colonne

Ogni CSV contiene:
- **101 colonne vol_***: Volumi cerebrali in mm³
  - Subcortical: hippocampus, amygdala, thalamus, caudate, putamen, pallidum, accumbens
  - Cortical: 68 regioni (Desikan-Killiany atlas)
  - Global: TIV, white matter, cortex, cerebellum, ventricles, CSF, brainstem
- **8 colonne qc_***: Quality control scores (0-1, higher is better)
  - qc_general white matter
  - qc_general grey matter
  - qc_general csf
  - qc_cerebellum
  - qc_brainstem
  - qc_thalamus
  - qc_putamen+pallidum
  - qc_hippocampus+amygdala
- **1 colonna subject_id**: Identificativo soggetto (ultima colonna)

